### PR TITLE
Backend Models

### PR DESCRIPTION
**Pull request by Dbz**
- [x] User
- [x] Review
- [x] Project
- [x] Bid

### CODE REVIEW

File: app/models/review.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+#  project_id  :integer          not null
+#  created_at  :datetime         not null
+#  updated_at  :datetime         not null
+#
+
+class Review < ActiveRecord::Base
+  validates :reviewer_id, presence: true
+  validates :reviewed_id, presence: true
+  validates :project_id, presence: true
+  validates :rating, presence: true
```
Comment (Dbz) [inline app/models/review.rb]: These can all be one line

File: app/models/project.rb [hunk 2]
```diff
@@ -7,6 +22,7 @@ class Project < ActiveRecord::Base
     :foreign_key => :user_id,
     :primary_key => :id
```
Comment (Dbz) [inline app/models/project.rb]: No need to list the primary key

File: app/models/review.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  validates :title, length: { minimum: 2 }
+  validates :body, length: { minimum: 2 }
```
Comment (Dbz) [inline app/models/review.rb]: These can be one line

File: .travis.yml [hunk 4]
```diff
@@ -1,3 +1,9 @@
 language: ruby
 rvm:
   - 2.2.1
+addons:
+  postgresql: "9.3"
+before_install:
+  - psql -c 'create database travic_ci_test;' -U postgres
```
Comment (Dbz) [inline .travis.yml]: Now that the CI builds, we should change this to a better database name

File: config/database.yml [hunk 5]
```diff
@@ -9,7 +9,7 @@ development:
   
 test:
   <<: *default
-  database: test_letspair
+  database: travis_ci_test
```
Comment (Dbz) [inline config/database.yml]: Change this back to test_letspair

**Comment by Dbz** [comment]
@isisAnchalee code review?

File: .travis.yml [hunk 6]
```diff
@@ -1,3 +1,9 @@
 language: ruby
 rvm:
   - 2.2.1
+addons:
+  postgresql: "9.3"
+before_install:
+  - psql -c 'create database test_letspair;' -U postgres
+before_script:
+  - rake db:reset
```
Comment (isisAnchalee) [inline .travis.yml]: newline

File: app/models/review.rb [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+#
+
+class Review < ActiveRecord::Base
+  validates :reviewer_id, :reviewed_id, :project_id, :rating, :title, :body, presence: true
+  validates :title, :body, length: { minimum: 2 }
+
+  belongs_to :reviewer, foreign_key: :reviewer_id, class_name: "User"
+  belongs_to :reviewed, foreign_key: :reviewed_id, class_name: "User"
+  belongs_to :project
+
```
Comment (isisAnchalee) [inline app/models/review.rb]: we should put a constraint on rating, I think we can pass it a regex matching proc to make it a valid number

Comment (isisAnchalee) [inline app/models/review.rb]: this should also be dependent -> destroy on the project

File: app/models/bid.rb [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+#  updated_at :datetime         not null
+#
+
+class Bid < ActiveRecord::Base
+  validates :project_id, :price, :bidder_id, :content, presence: true
+  validates :content, length: { minimum: 2 }
+  
+  belongs_to :project
+  belongs_to :bidder, foreign_key: :bidder_id, class_name: "User"
+
```
Comment (isisAnchalee) [inline app/models/bid.rb:20]: dependent -> destroy

File: spec/controllers/registrations_controller_spec.rb [hunk 9]
```diff
@@ -1,10 +1,10 @@
...
-end
\ No newline at end of file
+# describe "Users::RegistrationsController" do
+#   describe "#new" do
+#     it "should create a new user" do
+#       visit "/users/sign_up"
+#       expect(page).to have_selector "h2"
+#     end
+#   end
+# end
```
Comment (isisAnchalee) [inline spec/controllers/registrations_controller_spec.rb]: newline

File: spec/factories/bids.rb [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+#
+
+FactoryGirl.define do
+  factory :bid do
+    project_id 1
+    price 1
+    bidder_id 1
+    content "MyText"
+  end
+end
```
Comment (isisAnchalee) [inline spec/factories/bids.rb]: same

File: spec/factories/projects.rb [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+
+FactoryGirl.define do
+  factory :project do
+    title "Project Title"
+    description "This is a description"
+    complexity 10
+    price 10
+    user_id 1
+  end
+end
```
Comment (isisAnchalee) [inline spec/factories/projects.rb]: zzzzzzz

File: spec/factories/reviews.rb [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+FactoryGirl.define do
+  factory :review do
+    reviewer_id 1
+    reviewed_id 1
+    title "some title"
+    body "some text"
+    rating 10
+    project_id 1
+  end
+end
```
Comment (isisAnchalee) [inline spec/factories/reviews.rb]: ermagerrrrd

File: spec/factories/users.rb [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+    admin false
+  end
+  
+  factory :admin, class: User do
+    username "Admin"
+    password "hellohello"
+    admin true
+  end
+  
+end
```
Comment (isisAnchalee) [inline spec/factories/users.rb]: butts

File: spec/support/factory_girl.rb [hunk 14]
```diff
@@ -0,0 +1,3 @@
+RSpec.configure do |config|
+  config.include FactoryGirl::Syntax::Methods
+end
```
Comment (isisAnchalee) [inline spec/support/factory_girl.rb]: nl

File: app/models/review.rb (hunk 7 above)
Comment (Dbz) [inline app/models/review.rb]: I'll definitely add a validation for the rating. Do we want it in a range like 1-5?

Comment (Dbz) [inline app/models/review.rb]: I texted you about having a rating form. I'm thinking we won't want ratings to happen unless projects are completed.

File: app/models/review.rb [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+#  rating      :integer          not null
+#  project_id  :integer          not null
+#  created_at  :datetime         not null
+#  updated_at  :datetime         not null
+#
+
+class Review < ActiveRecord::Base
+  validates :reviewer_id, :reviewed_id, :project_id, :rating, :title, :body, presence: true
+  validates :title, :body, length: { minimum: 2 }
+  validates :rating, numericality: { greater_than_or_equal_to: 0, less_than_or_equal_to: 100 }
```
Comment (Dbz) [inline app/models/review.rb:19]: I added this validation, but I'm thinking we may change this in the future and create a rating model if we add a form for rating consultants

File: app/models/bid.rb (hunk 8 above)
Comment (Dbz) [inline app/models/bid.rb:20]: Added to project model

File: spec/models/project_spec.rb [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+    it "deletes its bids when destroyed" do
+      project = FactoryGirl.create(:project)
+      bid1 = FactoryGirl.create(:bid, project_id: project.id)
+      bid2 = FactoryGirl.create(:bid, project_id: project.id)
+      expect(Bid.all.count).to eq(2)
+      project.destroy
+      expect(Bid.all.count).to eq(0)
+    end
+  end
+end
```
Comment (isisAnchalee) [inline spec/models/project_spec.rb]: nl

File: app/models/review.rb (hunk 15 above)
Comment (isisAnchalee) [inline app/models/review.rb:19]: I don't think ratings need to be their own separate model, we will have a form for leaving a review right? The review's average rating generates the rating number?

Comment (Dbz) [inline app/models/review.rb:19]: If we have a separate detailed model it would give us the opportunity to have a nice visualization on their performance.

**Pull request merged**